File: Codes/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
/*
 * @Author: LetMeFly
 * @Date: 2025-03-13 12:52:32
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-03-13 12:59:04
 */
class Solution {
    private final char[] aeiou = {'a', 'e', 'i', 'o', 'u'};

    private int aeiouIndex(char c) {
        for (int i = 0; i < 5; i++) {
            if (aeiou[i] == c) {
                return i;
            }
        }
        return -1;
    }

    public long countOfSubstrings(String word, int k) {
        int[] bin1 = new int[5];
        int cntc1 = 0, cntk1 = 0, left1 = 0;
        int[] bin2 = new int[5];
        int cntc2 = 0, cntk2 = 0, left2 = 0;
        long ans = 0;
        for (char c : word.toCharArray()) {
            int index = aeiouIndex(c);
            if (index == -1) {
                cntk1++;
                cntk2++;
            } else {
                bin1[index]++;
                if (bin1[index] == 1) {
                    cntc1++;
                }
                bin2[index]++;
                if (bin2[index] == 1) {
                    cntc2++;
                }
            }

            while (cntc1 == 5 && cntk1 >= k) {
                index = aeiouIndex(word.charAt(left1++));
                if (index == -1) {
                    cntk1--;
                } else {
                    bin1[index]--;
                    if (bin1[index] == 0) {
                        cntc1--;
                    }
                }
            }
            while (cntc2 == 5 && cntk2 > k) {
                index = aeiouIndex(word.charAt(left2++));
                if (index == -1) {
                    cntk2--;
                } else {
                    bin2[index]--;
                    if (bin2[index] == 0) {
                        cntc2--;
                    }
                }
            }
            ans += left1 - left2;
        }
        return ans;
    }
}

File: Codes/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.go
/*
 * @Author: LetMeFly
 * @Date: 2025-03-13 12:59:35
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-03-13 13:04:36
 */
package main

var aeiou3306 []byte = []byte{'a', 'e', 'i', 'o', 'u'}

func aeiouIndex3306(c byte) int {
    for i := 0; i < 5; i++ {
        if (aeiou3306[i] == c) {
            return i
        }
    }
    return -1
}

func countOfSubstrings(word string, k int) (ans int64) {
    bin1 := make([]int, 5)
    cntc1, cntk1, left1 := 0, 0, 0
    bin2 := make([]int, 5)
    cntc2, cntk2, left2 := 0, 0, 0
    for i := range word {
        index := aeiouIndex3306(word[i])
        if index == -1 {
            cntk1++
            cntk2++
        } else {
            bin1[index]++
            if bin1[index] == 1 {
                cntc1++
            }
            bin2[index]++
            if bin2[index] == 1 {
                cntc2++
            }
        }

        for cntc1 == 5 && cntk1 >= k {
            index = aeiouIndex3306(word[left1])
            left1++
            if index == -1 {
                cntk1--
            } else {
                bin1[index]--
                if bin1[index] == 0 {
                    cntc1--
                }
            }
        }
        for cntc2 == 5 && cntk2 > k {
            index = aeiouIndex3306(word[left2])
            left2++
            if index == -1 {
                cntk2--
            } else {
                bin2[index]--
                if bin2[index] == 0 {
                    cntc2--
                }
            }
        }
        ans += int64(left1 - left2)
    }
    return
}